File: in_python/creational/Abstract_Factory_method/abstract_factory.py
from abc import ABC, abstractmethod
from enum import Enum

class Button(ABC):
    @abstractmethod
    def paint(self) -> str:
        pass

class Checkbox(ABC):
    @abstractmethod
    def paint(self) -> str:
        pass


#implementaciones concretas de los productos para Widows
class WinButton(Button):
    def paint(self) -> str:
        return "Rendering a button in Windows style"

class WinCheckbox(Checkbox):
    def paint(self) -> str:
        return "Rendering a checkbox in Windows style"
    
#implementaciones concretas de los productos para MacOS
class MacButton(Button):
    def paint(self) -> str:
        return "Rendering a button in MacOS style"

class MacCheckbox(Checkbox):
    def paint(self) -> str:
        return "Rendering a checkbox in MacOS style"


# --- Abstract Factory --- 
class GUIFactory(ABC):
    @abstractmethod
    def create_button(self) -> Button:
        pass

    @abstractmethod
    def create_checkbox(self) -> Checkbox:
        pass


# --fabricas concretas ---

class WinFactory(GUIFactory):
    def create_button(self) -> Button:
        return WinButton()

    def create_checkbox(self) -> Checkbox:
        return WinCheckbox()
    
class MacFactory(GUIFactory):
    def create_button(self) -> Button:
        return MacButton()

    def create_checkbox(self) -> Checkbox:
        return MacCheckbox()
    


class OSType(Enum):
    WINDOWS = "Windows"
    MACOS = "MacOS"


class GetGUIFactory:
    @staticmethod
    def get_factory(os_type: OSType) -> GUIFactory:
        if os_type == OSType.WINDOWS:
            return WinFactory()
        elif os_type == OSType.MACOS:
            return MacFactory()
        else:
            raise ValueError(f"Unknown OS type: {os_type}") 

File: in_rust/creational/src/abstract_factory_method/abstract_factory.rs
#![allow(dead_code)]

pub trait Button {
    fn render(&self) -> String;
}
pub trait Checkbox {
    fn render(&self) -> String;
}

//------------------- Concrete Products for windows ------------------
struct WindowsButton;
impl Button for WindowsButton {
    fn render(&self) -> String {
        "Rendering Windows Button".to_string()
    }
}

struct WindowsCheckbox;
impl Checkbox for WindowsCheckbox {
    fn render(&self) -> String {
        "Rendering Windows Checkbox".to_string()
    }
}

//------------------- Concrete Products for macOS ------------------
struct MacOSButton;
impl Button for MacOSButton {
    fn render(&self) -> String {
        "Rendering macOS Button".to_string()
    }
}

struct MacOSCheckbox;
impl Checkbox for MacOSCheckbox {
    fn render(&self) -> String {
        "Rendering macOS Checkbox".to_string()
    }
}
//------------------- Abstract Factory ------------------
pub trait GUIFactory {
    fn create_button(&self) -> Box<dyn Button>;
    fn create_checkbox(&self) -> Box<dyn Checkbox>;
}

//------------------- Concrete Factories ------------------
struct WindowsFactory;
impl GUIFactory for WindowsFactory {
    fn create_button(&self) -> Box<dyn Button> {
        Box::new(WindowsButton)
    }

    fn create_checkbox(&self) -> Box<dyn Checkbox> {
        Box::new(WindowsCheckbox)
    }
}

struct MacOSFactory;
impl GUIFactory for MacOSFactory {
    fn create_button(&self) -> Box<dyn Button> {
        Box::new(MacOSButton)
    }
    fn create_checkbox(&self) -> Box<dyn Checkbox> {
        Box::new(MacOSCheckbox)
    }
}

pub enum OSType {
    Windows,
    Macos,
}

//------------------- Client Code ------------------
pub struct GetGuiFactory;
impl GetGuiFactory {
    pub fn get_factory(os_type: OSType) -> Box<dyn GUIFactory> {
        match os_type {
            OSType::Windows => Box::new(WindowsFactory),
            OSType::Macos => Box::new(MacOSFactory),
        }
    }
}
